modules/bezug_rct2/rct_read_speicher.py

```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/python3
+import os
+import sys
+import rct_lib
+import fnmatch
+#
+# Author Heinz Hoefling
+# Version 1.0 Okt.2021
+# Fragt die Werte gebuendelt ab, 
+#
+
+def writeRam(fn,val, rctname):
+    fnn = "/var/www/html/openWB/ramdisk/"+str(fn)
+    if rct_lib.bVerbose == True:
+        f = open(fnn,'r')
+        oldv = f.read()
+        f.close()
+        rct_lib.dbglog("field " + str(fnn)+ " val is "+ str(val) + " oldval:"+ str(oldv) + " "  + str(rctname) )
+    
+    f = open(fnn,'w')
+    f.write(str(val))
+    f.close()
+    
+
+# Entry point with parameter check
+def main():
+    rct_lib.init(sys.argv)
+
+    clientsocket = rct_lib.connect_to_server()
+    if clientsocket is not None:
+
+        socx = rct_lib.read(clientsocket,0x959930BF )
+        soc =int(socx * 100.0)
+        writeRam('speichersoc', soc, '0x959930BF battery.soc')
+
+        watt =int(rct_lib.read(clientsocket,0x400F015B ) * -1.0 ) 
+        writeRam('speicherleistung', watt, '0x400F015B g_sync.p_acc_lp')
+
+        watt =int(rct_lib.read(clientsocket,0x5570401B ))
+        #rct_lib.dbglog("speicherikwh will be battery.stored_energy "+ str(watt)) 
+        writeRam('speicherikwh', watt, '0x5570401B battery.stored_energy')
+
+        watt =int(rct_lib.read(clientsocket,0xA9033880 ))
+        #rct_lib.dbglog("speicherekwh will be battery.used_energy "+ str(watt))         
+        writeRam('speicherekwh', watt, '#0xA9033880 battery.used_energy')
+
+        stat1 = int(rct_lib.read(clientsocket,0x70A2AF4F ))
+        rct_lib.dbglog("battery.bat_status "+ str(stat1))
+
+        stat2 = int(rct_lib.read(clientsocket,0x71765BD8 ))
+        rct_lib.dbglog("battery.status "+ str(stat2))
+
+        stat3 = int(rct_lib.read(clientsocket,0x0DE3D20D ))
+        rct_lib.dbglog("battery.status2 "+ str(stat3))
+        
+        faultStr=''
+        faultState=0
+
+        if ( stat1 + stat2 + stat3) > 0:
+            faultStr = "Battery ALARM Battery-Status nicht 0"
+            faultState=2
+             # speicher in mqtt 
+           
+        os.system('mosquitto_pub -r -t openWB/housebattery/faultState -m "' + str(faultState) +'"')
+        os.system('mosquitto_pub -r -t openWB/housebattery/faultStr -m "' + str(faultStr) +'"')
+        
+        socsoll = int(rct_lib.read(clientsocket,0x8B9FF008 ) * 100.0 )
+        os.system('mosquitto_pub -r -t openWB/housebattery/soctarget -m "' + str(socsoll) +'"')
```

Comment:
Die Parameter an `faultState`, `faultStr` und `socsoll` werden so nicht escaped. Wenn einer der Parameter ein Anführungszeichen enthält geht das alles den Bach runter. Aktuell ist das zwar nicht der Fall, aber irgendwann editiert jemand die Datei und ändert was und plötzlich kann auch ein Anführungszeichen drin vorkommen.

Außerdem ruft `os.system` einfach den "command processor" der "host environment" auf. Es ist nicht so wirklich spezifiziert welcher das ist und wie das Kommando genau ausgewertet wird. Bei einfachen Kommandos mag das alles gut gehen, weil die dann doch ähnlich funktionieren, aber wenn es nicht nötig ist, würde ich das vermeiden.

Die Doku zu [os.system](https://docs.python.org/3/library/os.html#os.system) sie das ähnlich:
> The subprocess module provides more powerful facilities for spawning new processes and retrieving their results; using that module is preferable to using this function.

Damit geht es dann auch sehr gut:
```python
from subprocess import call
call(["mosquitto_pub", "-r", "-t", "openWB/housebattery/soctarget", "-m", str(socsoll)])
```
Dann läuft das ganze systemunabhängig und ohne sich den Kopf über escaping zu zerbrechen.

Noch cooler ist es natürlich das ganze gleich in Python zu machen. Etwa so:
```python
import paho.mqtt.client as mqtt

client = mqtt.Client("openWB-python-rct-" + str(os.getpid()))
client.connect("localhost", 1886)
client.publish("openWB/housebattery/soctarget", str(socsoll), qos=0, retain=True)
# ...
client.disconnect()
```
Ungetestet...

Theoretisch gibt es seit neuestem in `packages/helpermodules/pub.py` Hilfsfunktionen dafür. Aber ob die schon einsatzbereit sind... Weiß nicht...

Reply:
Danke für die vielen Optimierenansätze.
Aber das ist meine erste Python Werk.
Vieles ist von anderen Script abkopiert. 
Es tut seine Arbeit (bei mir) darf aber gerne Weiterentwickelt werden.

Reply:
Ja da könnte man noch weiter optimiern.
Code wurde von anderen Modulen kopiert.

Reply:
> Danke für die vielen Optimierenansätze. Aber das ist meine erste Python Werk. Vieles ist von anderen Script abkopiert. Es tut seine Arbeit (bei mir) darf aber gerne Weiterentwickelt werden.

Zwischen den Zeilen habe ich den Eindruck, dass dich meine ganzen Kommentare demotivieren. Das tut mir Leid, das war nicht meine Absicht.

Das "Kopieren aus anderen Modulen" ist leider ein Teufelskreis. Viele Module hier haben das Problem, dass sie Code haben der zwar irgendwie funktioniert, aber schwierig zu verstehen ist und/oder der fragil ist und unter bestimmten Umständen kaputt gehen kann. Wenn dieser Code immer wieder kopiert wird, wird das ganze natürlich nicht besser. Ich würde diesen Teufelskreis gerne durchbrechen. Daher habe ich hier alles mal kommentiert, was mir so aufgefallen ist.

Es tut mir Leid, dass dich das jetzt getroffen hat. Aber wie gesagt: Ich bin niemand offizielles. Ich habe keinerlei Entscheidungsmacht darüber, was mit dem PR hier passiert. Wenn du Lust hast, lade ich dich ein das beste und strukturierteste Modul zu bauen, was oWB je gesehen hat. Aber es ist nur eine Einladung von einem Außenstehenden. Nichts daran ist bindend.

Reply:
Dein Eindruck trügt dich nicht.
Aber ich sehe das eher Pragmatisch.
Das ganze openWB ist in der 1.9 sowieso eher "wie gewachsen" bzw. "real live"
Sollange aber die Mitarbeit hier so stieifmütterlich gemanaged wird (wohl auch, hoffentlich, weil die 2.0 Vorrang hat)
Werde ich da nichts weiter machen. Wenn das Module es endlich in die 1.9'er geschaft hat, können wir ja  gemeinsame Detailpflege betreiben. Ansonsten rausgeschmissene Zeit.
Wobei, vieles was du Anmerkst unterliegt auch dem persönlichen Gescmack.